### DISCUSSION TITLE
Next Completion With Tab, Completion Menu With A Keybind

### DISCUSSION THREAD

**Discussion by CamRatliff** [akinomyoga/ble.sh#439] [Q&A] [answered]
I prefer to see the interactive completion menu only if I ask for it with a keybind. For example, when I press TAB I like it to just fill the first completion, press tab again the completion after that, and so on. If I want to see the completion menu, I like to press CTRL+SPACE.

I'm able to sort of make this work in plain BASH, with out the interactive menu of course, by using this in my bashrc:
```
bind '"\t":menu-complete' # tab
bind '"\e[Z":menu-complete-backward' # shift+tab
bind '"\C-@":possible-completions' # ctrl+space
```
I was able to get this to work in zsh with this:

```
# SHIFT+TAB to reverse menu complete
bindkey '^[[Z' reverse-menu-complete

# Auto complete on first press of TAB
setopt MENU_COMPLETE
# Don't expand the completion menu upon pressing TAB
unsetopt AUTO_LIST

# Create a completion function for an interactive menu and bind it to CTRL+SPACE
zle -C menu-select-interactive menu-complete _generic
zstyle ':completion:menu-select-interactive:*' menu select interactive
bindkey '^ ' menu-select-interactive
```
I've looked through configs and settings, but I couldn't find what I'm looking for. Is there a way to duplicate this set up with BLE.SH?

**Comment by akinomyoga** [comment]
Thank you. I haven't thought about that usage. Currently ble.sh doesn't have the feature to go through the ~~history~~ menu items without showing the menu. I'll add it in the todo list.

**Reply by CamRatliff** [reply]
Sick, thanks

**Reply by akinomyoga** [reply, marked as answer]
@CamRatliff I added `bleopt complete_menu_complete_opts=hidden` in commit 16ff7df7. The visibility of the menu can be toggled by <kbd>C-x SP</kbd> in the menu-complete mode. I also added the feature to switch the menu style in the menu-complete mode by <kbd>C-x a</kbd>, <kbd>C-x c</kbd>, <kbd>C-x d</kbd>, <kbd>C-x l</kbd>, <kbd>C-x left</kbd> and <kbd>C-x right</kbd> (see the table in [§7.2](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A77-Completion#user-content-sec-menu-complete) for details).

To get a similar behavior as your configuration for plain Bash/Zsh, please put the following setting in your `~/.bashrc`. (Note that depending on the terminal TAB and C-SP may be transmitted as C-i and C-@).

```bash
# blerc
ble-import core-complete -C '
  bleopt complete_menu_complete_opts=insert-selection:hidden
  ble-bind -f TAB menu-complete
  ble-bind -f C-i menu-complete
  ble-bind -f C-@  "complete toggle_menu"
  ble-bind -f C-SP "complete toggle_menu"
  ble-bind -m menu_complete -f C-@  menu_complete/toggle-hidden
  ble-bind -m menu_complete -f C-SP menu_complete/toggle-hidden
'
```

Could you check the behavior?

**Reply by CamRatliff** [reply]
Works great. Thanks so much

**Reply by akinomyoga** [reply]
Thank you for checking!
